### DISCUSSION TITLE
Sync with CPU profiling

### DISCUSSION THREAD

**Discussion by gents83** [Wumpf/wgpu-profiler#49] [Ideas]
Do you think it would be possible to have some kind of parameter per-frame to adjust somehow timings in output from wgpu-profiler to the one of a cpu profiler?

As example I've bound a chrome tracing profilier for cpu and I'm using wgpu-profiler for gpu.

It would be awesome to show them in the same tracing,
so offsetting\shifting the gpu ones to match the start of cpu ones in the frame.

I know it'll be never representative of the reality but it would be anyway nice to see them in the same "frame" and with their own correct timings (in millisec or microsec)

**Comment by Wumpf** [comment]
I had this idea of adding cpu timestamps to [`GpuTimerScopeResult`](https://github.com/Wumpf/wgpu-profiler/blob/main/src/lib.rs#L91) so you know when a thing was recorded that the gpu then processes later. Giving you a sense of "the cpu spend this much time of this pass, the gpu that much".
Would that do what you want? That's not doing any of the syncing up, but I don't think there's a meaningful way to do that.

**Reply by gents83** [reply]
That would be useful too.
What I had in mind was more to be "pass the current CPU time" during the "begin_scope" and\or the "end_scope" 
and then "add it" as an offset to the GPU timings
so that in a chrometrace with both CPU and GPU profiling they will be shown in the "same area".
You see what I mean?

**Reply by Wumpf** [reply]
if its all about aligning the gpu and cpu frames in the profiler view then I fail to see how it's useful to do any kind of offset per scope. What you really want then is track which gpu frames belong to which cpu frames and offset according to that?

**Reply by Wumpf** [reply]
Can you maybe do a quick pseudo-code scribble to illustrate what it would look like?

**Reply by gents83** [reply]
Syncing it with the frame as well would do the job probably, yes.
What I was thinking was more to be able to store in the begin_scope() the current cpu time or the frame time or the time from the start of the application and then being able to offset the gpu timings of that value, in order to have them almost aligned with cpu ones.
Doing that once per frame or at each begin_scope probably will achieve same result, because we cannot anyway be precise.

**Reply by Wumpf** [reply]
The expectation is that you have lots and lots of `begin_scope` calls, so storing cpu time for the purpose of aligning gpu and cpu time seems quite odd to me. Buuut I think it's potentially useful regardless as posted initially. So I'd be open to the idea of putting that optionally (!) in. Also worth considering to instead allow for arbitrary (`Option<Box<std::any::Any>>`, or alternatively add a user-data-type type to GpuProfiler -> `Option<Box<UserDataType>>`) user payload 🤔

**Reply by Imberflur** [reply]
I just encountered a case where it would be useful to at least of some sort of identifier for a frame so that other metrics collected for that frame can be easily associated to the GPU timings.

**Reply by Wumpf** [reply]
let's have custom payloads on both scopes and profiler frames, otherwise more of these will keep coming up 🤔

**Reply by Wumpf** [reply]
fyi another very big bundle of api changes if coming soon anyways, making GpuProfiler threadsafe. So might make sense to bundle all these into a release:
* https://github.com/Wumpf/wgpu-profiler/issues/46
* https://github.com/Wumpf/wgpu-profiler/issues/51
* https://github.com/Wumpf/wgpu-profiler/issues/54

**Reply by Wumpf** [reply]
I think being able to add arbitrary data to scopes & frames would solve your syncing ask as well @gents83 , but wanted to check if that's too far off from what you had in mind?

**Reply by gents83** [reply]
Absolutely! I think that once that we've data that we can pass and then retrieve at the end to do our logic - we can do whatever then :) Tnx a lot @Wumpf
